File: CPI_Forecasting_Project/cpi_forecasting.R
# Forecast Malaysia's low-income household CPI using four time-series methods.
# Run from this folder in Posit Cloud with: Rscript cpi_forecasting.R

required_packages <- c("readr", "dplyr", "ggplot2", "forecast")
missing_packages <- required_packages[!vapply(required_packages, requireNamespace, logical(1), quietly = TRUE)]
if (length(missing_packages) > 0) {
  stop(
    paste0(
      "Missing package(s): ", paste(missing_packages, collapse = ", "),
      ". Install them with: install.packages(c(",
      paste(sprintf("'%s'", missing_packages), collapse = ", "), "))"
    ),
    call. = FALSE
  )
}

suppressPackageStartupMessages({
  library(readr)
  library(dplyr)
  library(ggplot2)
  library(forecast)
})

project_dir <- normalizePath(getwd(), winslash = "/", mustWork = TRUE)
data_file <- file.path(project_dir, "cpi_2d_lowincome.csv")
figures_dir <- file.path(project_dir, "outputs", "figures")
tables_dir <- file.path(project_dir, "outputs", "tables")
forecast_horizon <- 12
seasonal_period <- 12

dir.create(figures_dir, recursive = TRUE, showWarnings = FALSE)
dir.create(tables_dir, recursive = TRUE, showWarnings = FALSE)

save_plot <- function(plot_object, filename, width = 11, height = 5) {
  ggsave(
    filename = file.path(figures_dir, filename), plot = plot_object,
    width = width, height = height, units = "in", dpi = 300, bg = "white"
  )
}

fail <- function(message) {
  stop(paste("DATA VALIDATION FAILED:", message), call. = FALSE)
}

make_monthly_ts <- function(values, dates) {
  ts(values, start = c(as.integer(format(min(dates), "%Y")), as.integer(format(min(dates), "%m"))), frequency = 12)
}

metric_values <- function(actual, predicted) {
  errors <- actual - predicted
  tibble(
    MAE = mean(abs(errors)),
    RMSE = sqrt(mean(errors^2)),
    `MAPE (%)` = mean(abs(errors / actual)) * 100
  )
}

if (!file.exists(data_file)) {
  fail("Cannot find cpi_2d_lowincome.csv beside cpi_forecasting.R.")
}

data <- read_csv(data_file, show_col_types = FALSE)
cat("Forecasting Malaysia's Monthly Consumer Price Index for Low-Income Households\n\n")
cat("--- Raw dataset preview ---\n")
cat("Shape:", nrow(data), "rows and", ncol(data), "columns\n")
cat("Columns:", paste(names(data), collapse = ", "), "\n")
cat("Data types before conversion:\n")
print(vapply(data, class, character(1)))
cat("First five rows:\n")
print(head(data, 5))

required_columns <- c("date", "division", "index")
absent_columns <- setdiff(required_columns, names(data))
if (length(absent_columns) > 0) fail(paste("Essential column(s) missing:", paste(absent_columns, collapse = ", ")))

data <- data %>% mutate(date = as.Date(date), index = suppressWarnings(as.numeric(index)))
if (any(is.na(data$date))) fail("The date column contains values that cannot be converted to dates.")
if (any(is.na(data$index))) fail("The index column contains values that cannot be converted to numeric values.")

missing_counts <- colSums(is.na(data))
duplicate_records <- sum(duplicated(data[c("date", "division")]))
if (any(missing_counts > 0)) fail(paste("Missing values found in:", paste(names(missing_counts)[missing_counts > 0], collapse = ", ")))
if (duplicate_records > 0) fail(paste("Found", duplicate_records, "duplicate date-division record(s)."))

divisions <- sort(unique(data$division))
if (!("overall" %in% divisions)) fail("The required division 'overall' is not available.")
overall <- data %>% filter(division == "overall") %>% arrange(date)
if (nrow(overall) <= forecast_horizon + seasonal_period) fail("The overall series is too short for a seasonal model and 12-month test set.")

expected_dates <- seq(min(overall$date), max(overall$date), by = "month")
missing_months <- setdiff(expected_dates, overall$date)
if (length(missing_months) > 0) fail(paste("Missing month(s) in overall CPI:", paste(format(missing_months, "%Y-%m"), collapse = ", ")))

cat("\n--- Validation summary ---\n")
cat("Missing values:", sum(missing_counts), "\n")
cat("Duplicate date-division combinations:", duplicate_records, "\n")
cat("Available divisions (", length(divisions), "): ", paste(divisions, collapse = ", "), "\n", sep = "")
cat("Overall CPI observations:", nrow(overall), "\n")
cat("Overall CPI date range:", format(min(overall$date), "%B %Y"), "to", format(max(overall$date), "%B %Y"), "\n")
cat("Overall CPI range:", sprintf("%.1f", min(overall$index)), "to", sprintf("%.1f", max(overall$index)), "\n")
cat("Missing months in overall CPI:", length(missing_months), "\n")
cat("Result: all essential validation checks passed.\n")

prepared <- overall %>%
  mutate(
    monthly_change_pct = (index / lag(index) - 1) * 100,
    yoy_inflation_pct = (index / lag(index, 12) - 1) * 100,
    year = as.integer(format(date, "%Y")),
    month_number = as.integer(format(date, "%m")),
    month_name = format(date, "%B")
  )
write_csv(prepared, file.path(tables_dir, "cleaned_overall_cpi.csv"), na = "")

descriptive_statistics <- bind_rows(lapply(c("index", "monthly_change_pct", "yoy_inflation_pct"), function(variable) {
  values <- prepared[[variable]]
  tibble(
    variable = variable, count = sum(!is.na(values)), mean = mean(values, na.rm = TRUE),
    sd = sd(values, na.rm = TRUE), min = min(values, na.rm = TRUE),
    q1 = quantile(values, 0.25, na.rm = TRUE), median = median(values, na.rm = TRUE),
    q3 = quantile(values, 0.75, na.rm = TRUE), max = max(values, na.rm = TRUE)
  )
}))
write_csv(descriptive_statistics, file.path(tables_dir, "descriptive_statistics.csv"), na = "")

latest <- prepared[nrow(prepared), ]
max_yoy <- prepared %>% filter(yoy_inflation_pct == max(yoy_inflation_pct, na.rm = TRUE)) %>% slice(1)
min_yoy <- prepared %>% filter(yoy_inflation_pct == min(yoy_inflation_pct, na.rm = TRUE)) %>% slice(1)
cat("\n--- Factual descriptive interpretations ---\n")
cat("Latest CPI is", sprintf("%.1f", latest$index), "in", format(latest$date, "%B %Y"), ".\n")
cat("Latest monthly change is", sprintf("%.2f%%", latest$monthly_change_pct), "and latest year-on-year inflation is", sprintf("%.2f%%", latest$yoy_inflation_pct), ".\n")
cat("Highest year-on-year inflation is", sprintf("%.2f%%", max_yoy$yoy_inflation_pct), "in", format(max_yoy$date, "%B %Y"), ".\n")
cat("Lowest year-on-year inflation is", sprintf("%.2f%%", min_yoy$yoy_inflation_pct), "in", format(min_yoy$date, "%B %Y"), ".\n")

base_theme <- theme_minimal(base_size = 12) + theme(panel.background = element_rect(fill = "white", colour = NA))
save_plot(ggplot(prepared, aes(date, index)) + geom_line(colour = "#1f5a99", linewidth = 0.8) + labs(title = "Malaysia Low-Income Household Overall CPI", x = "Date", y = "CPI index") + base_theme, "01_overall_cpi_line.png")
save_plot(ggplot(prepared, aes(date, monthly_change_pct)) + geom_line(colour = "#d95f02", linewidth = 0.7) + geom_hline(yintercept = 0) + labs(title = "Monthly Percentage Change in Overall CPI", x = "Date", y = "Monthly change (%)") + base_theme, "02_monthly_percentage_change.png")
save_plot(ggplot(prepared, aes(date, yoy_inflation_pct)) + geom_line(colour = "#7570b3", linewidth = 0.7) + geom_hline(yintercept = 0) + labs(title = "Year-on-Year Inflation in Overall CPI", x = "Date", y = "Year-on-year inflation (%)") + base_theme, "03_yoy_inflation.png")
save_plot(ggplot(prepared, aes(factor(month_number, levels = 1:12, labels = month.abb), index)) + geom_boxplot(fill = "#8dd3c7") + labs(title = "Seasonal Distribution of Overall CPI by Calendar Month", x = "Calendar month", y = "CPI index") + base_theme, "04_seasonal_cpi_boxplot.png")

overall_ts <- make_monthly_ts(prepared$index, prepared$date)
decomposition <- decompose(overall_ts, type = "additive")
png(file.path(figures_dir, "05_seasonal_trend_decomposition.png"), width = 3300, height = 2400, res = 300)
plot(decomposition, main = "Additive Seasonal-Trend Decomposition of Overall CPI")
dev.off()
png(file.path(figures_dir, "06_acf.png"), width = 3000, height = 1500, res = 300)
Acf(overall_ts, lag.max = 24, main = "Autocorrelation Function of Overall CPI", xlab = "Lag (months)")
dev.off()
png(file.path(figures_dir, "07_pacf.png"), width = 3000, height = 1500, res = 300)
Pacf(overall_ts, lag.max = 24, main = "Partial Autocorrelation Function of Overall CPI", xlab = "Lag (months)")
dev.off()

training <- prepared %>% slice_head(n = n() - forecast_horizon)
testing <- prepared %>% slice_tail(n = forecast_horizon)
write_csv(training, file.path(tables_dir, "training_data.csv"), na = "")
write_csv(testing, file.path(tables_dir, "testing_data.csv"), na = "")
cat("\n--- Chronological train-test split ---\n")
cat("Training:", nrow(training), "observations,", format(min(training$date), "%Y-%m"), "to", format(max(training$date), "%Y-%m"), "\n")
cat("Testing:", nrow(testing), "observations,", format(min(testing$date), "%Y-%m"), "to", format(max(testing$date), "%Y-%m"), "\n")
split_plot_data <- bind_rows(training %>% transmute(date, index, period = "Training"), testing %>% transmute(date, index, period = "Testing"))
save_plot(ggplot(split_plot_data, aes(date, index, colour = period)) + geom_line(linewidth = 0.8) + geom_vline(xintercept = as.numeric(min(testing$date)), linetype = "dashed") + scale_colour_manual(values = c("Training" = "#1f5a99", "Testing" = "#e7298a")) + labs(title = "Chronological Training and Testing Periods", x = "Date", y = "CPI index", colour = NULL) + base_theme, "08_training_testing_split.png")

train_ts <- make_monthly_ts(training$index, training$date)
seasonal_naive_fc <- snaive(train_ts, h = forecast_horizon)
holt_winters_fc <- hw(train_ts, seasonal = "additive", h = forecast_horizon)
sarima_model <- Arima(train_ts, order = c(1, 1, 1), seasonal = c(1, 1, 1))
sarima_fc <- forecast(sarima_model, h = forecast_horizon)
regression_model <- tslm(train_ts ~ trend + season)
regression_fc <- forecast(regression_model, h = forecast_horizon)

forecasts <- list(
  "Seasonal Naive" = as.numeric(seasonal_naive_fc$mean),
  "Holt-Winters" = as.numeric(holt_winters_fc$mean),
  "SARIMA" = as.numeric(sarima_fc$mean),
  "Trend + Monthly Dummies" = as.numeric(regression_fc$mean)
)
forecast_file_names <- c(
  "Seasonal Naive" = "seasonal_naive",
  "Holt-Winters" = "holt_winters",
  "SARIMA" = "sarima",
  "Trend + Monthly Dummies" = "trend_plus_monthly_dummies"
)
results <- testing %>% transmute(date, actual_cpi = index)
metrics <- tibble()
for (model_name in names(forecasts)) {
  prediction <- forecasts[[model_name]]
  safe_name <- forecast_file_names[[model_name]]
  results[[safe_name]] <- prediction
  metrics <- bind_rows(metrics, bind_cols(tibble(model = model_name), metric_values(testing$index, prediction)))
  plot_data <- bind_rows(
    training %>% transmute(date, index, series = "Training"),
    testing %>% transmute(date, index, series = "Actual test CPI"),
    tibble(date = testing$date, index = prediction, series = paste(model_name, "forecast"))
  )
  forecast_label <- paste(model_name, "forecast")
  colours <- c("Training" = "#bdbdbd", "Actual test CPI" = "#e7298a")
  colours[forecast_label] <- "#1f5a99"
  save_plot(ggplot(plot_data, aes(date, index, colour = series, linetype = series)) + geom_line(linewidth = 0.8) + geom_point(data = filter(plot_data, series != "Training"), size = 1.6) + scale_colour_manual(values = colours) + labs(title = paste(model_name, ": 12-Month CPI Forecast"), x = "Date", y = "CPI index", colour = NULL, linetype = NULL) + base_theme, paste0("09_forecast_", safe_name, ".png"))
}
metrics <- metrics %>% arrange(RMSE)
write_csv(results, file.path(tables_dir, "test_forecasts.csv"))
write_csv(metrics, file.path(tables_dir, "forecast_accuracy_metrics.csv"))
cat("\n--- Forecast accuracy on the final 12 months ---\n")
print(metrics)
cat("Lowest RMSE model:", metrics$model[1], "(", sprintf("%.3f", metrics$RMSE[1]), ").\n", sep = "")
save_plot(ggplot(metrics, aes(reorder(model, RMSE), RMSE, fill = model)) + geom_col(show.legend = FALSE) + coord_flip() + labs(title = "Forecast Accuracy Comparison: Lower RMSE Is Better", x = "Forecasting model", y = "RMSE (CPI index points)") + base_theme, "10_forecast_accuracy_comparison.png")

cat("\nCompleted successfully. Tables and figures are in outputs/.\n")

File: CPI_Forecasting_Project/cpi_forecasting.py
"""Forecast Malaysia's low-income household CPI using four time-series methods.

Run from this folder with: python3 cpi_forecasting.py
"""

from pathlib import Path
import sys
import warnings

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
from sklearn.linear_model import LinearRegression
from statsmodels.graphics.tsaplots import plot_acf, plot_pacf
from statsmodels.tsa.holtwinters import ExponentialSmoothing
from statsmodels.tsa.seasonal import seasonal_decompose
from statsmodels.tsa.statespace.sarimax import SARIMAX


PROJECT_DIR = Path(__file__).resolve().parent
DATA_FILE = PROJECT_DIR / "cpi_2d_lowincome.csv"
FIGURES_DIR = PROJECT_DIR / "outputs" / "figures"
TABLES_DIR = PROJECT_DIR / "outputs" / "tables"
FORECAST_HORIZON = 12
SEASONAL_PERIOD = 12
COLOUR = "#1f5a99"
MONTH_ORDER = list(range(1, 13))
MONTH_LABELS = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]


def setup_output_folders():
    """Create the output folders when they are not already present."""
    FIGURES_DIR.mkdir(parents=True, exist_ok=True)
    TABLES_DIR.mkdir(parents=True, exist_ok=True)


def save_figure(filename):
    """Apply consistent figure saving and release the figure from memory."""
    plt.tight_layout()
    plt.savefig(FIGURES_DIR / filename, dpi=300, bbox_inches="tight")
    plt.close()


def fail(message):
    """Stop the program with an actionable validation message."""
    raise ValueError(f"DATA VALIDATION FAILED: {message}")


def load_and_validate_data():
    """Load the CSV and validate the structure needed for forecasting."""
    if not DATA_FILE.exists():
        fail(f"Cannot find '{DATA_FILE.name}'. Place it beside cpi_forecasting.py and run again.")

    data = pd.read_csv(DATA_FILE)
    print("\n--- Raw dataset preview ---")
    print(f"Shape: {data.shape}")
    print(f"Columns: {data.columns.tolist()}")
    print("Data types before conversion:")
    print(data.dtypes)
    print("First five rows:")
    print(data.head().to_string(index=False))

    required_columns = {"date", "division", "index"}
    missing_columns = required_columns.difference(data.columns)
    if missing_columns:
        fail(f"Essential column(s) missing: {sorted(missing_columns)}")

    try:
        data["date"] = pd.to_datetime(data["date"], errors="raise")
    except (ValueError, TypeError) as error:
        fail(f"The date column cannot be converted to dates: {error}")
    try:
        data["index"] = pd.to_numeric(data["index"], errors="raise")
    except (ValueError, TypeError) as error:
        fail(f"The index column is not numeric: {error}")

    missing_values = data.isna().sum()
    duplicate_records = data.duplicated(subset=["date", "division"]).sum()
    if missing_values.any():
        fail(f"Missing values found: {missing_values[missing_values > 0].to_dict()}")
    if duplicate_records:
        fail(f"Found {duplicate_records} duplicate date-division record(s).")

    divisions = sorted(data["division"].astype(str).unique())
    if "overall" not in divisions:
        fail("The required division 'overall' is not available.")

    overall = data.loc[data["division"] == "overall"].copy().sort_values("date")
    if overall.empty:
        fail("The overall CPI series is empty.")
    expected_dates = pd.date_range(overall["date"].min(), overall["date"].max(), freq="MS")
    missing_months = expected_dates.difference(pd.DatetimeIndex(overall["date"]))
    if len(missing_months):
        fail("Missing month(s) in overall CPI: " + ", ".join(date.strftime("%Y-%m") for date in missing_months))
    if len(overall) <= FORECAST_HORIZON + SEASONAL_PERIOD:
        fail("The overall series is too short for a seasonal model and 12-month test set.")

    print("\n--- Validation summary ---")
    print(f"Missing values: {int(missing_values.sum())}")
    print(f"Duplicate date-division combinations: {duplicate_records}")
    print(f"Available divisions ({len(divisions)}): {divisions}")
    print(f"Overall CPI observations: {len(overall)}")
    print(f"Overall CPI date range: {overall['date'].min():%B %Y} to {overall['date'].max():%B %Y}")
    print(f"Overall CPI range: {overall['index'].min():.1f} to {overall['index'].max():.1f}")
    print(f"Missing months in overall CPI: {len(missing_months)}")
    print("Result: all essential validation checks passed.")
    return data, overall


def prepare_data(overall):
    """Add descriptive time-series variables without changing the original CSV."""
    prepared = overall.copy()
    prepared["monthly_change_pct"] = prepared["index"].pct_change() * 100
    prepared["yoy_inflation_pct"] = prepared["index"].pct_change(12) * 100
    prepared["year"] = prepared["date"].dt.year
    prepared["month_number"] = prepared["date"].dt.month
    prepared["month_name"] = prepared["date"].dt.month_name()
    prepared.to_csv(TABLES_DIR / "cleaned_overall_cpi.csv", index=False)

    descriptive = prepared[["index", "monthly_change_pct", "yoy_inflation_pct"]].describe().T
    descriptive.to_csv(TABLES_DIR / "descriptive_statistics.csv")
    return prepared


def print_interpretations(prepared):
    """Print short interpretations calculated from the current data."""
    latest = prepared.iloc[-1]
    max_yoy = prepared.loc[prepared["yoy_inflation_pct"].idxmax()]
    min_yoy = prepared.loc[prepared["yoy_inflation_pct"].idxmin()]
    print("\n--- Factual descriptive interpretations ---")
    print(f"Latest CPI is {latest['index']:.1f} in {latest['date']:%B %Y}.")
    print(f"Latest monthly change is {latest['monthly_change_pct']:.2f}% and latest year-on-year inflation is {latest['yoy_inflation_pct']:.2f}%.")
    print(f"Highest year-on-year inflation is {max_yoy['yoy_inflation_pct']:.2f}% in {max_yoy['date']:%B %Y}.")
    print(f"Lowest year-on-year inflation is {min_yoy['yoy_inflation_pct']:.2f}% in {min_yoy['date']:%B %Y}.")


def create_exploratory_figures(prepared):
    """Create the required descriptive and diagnostic figures."""
    sns.set_theme(style="whitegrid", context="notebook")

    plt.figure(figsize=(11, 5))
    plt.plot(prepared["date"], prepared["index"], color=COLOUR, linewidth=2)
    plt.title("Malaysia Low-Income Household Overall CPI")
    plt.xlabel("Date")
    plt.ylabel("CPI index")
    plt.xticks(rotation=35)
    save_figure("01_overall_cpi_line.png")

    plt.figure(figsize=(11, 5))
    plt.plot(prepared["date"], prepared["monthly_change_pct"], color="#d95f02", linewidth=1.5)
    plt.axhline(0, color="black", linewidth=0.8)
    plt.title("Monthly Percentage Change in Overall CPI")
    plt.xlabel("Date")
    plt.ylabel("Monthly change (%)")
    plt.xticks(rotation=35)
    save_figure("02_monthly_percentage_change.png")

    plt.figure(figsize=(11, 5))
    plt.plot(prepared["date"], prepared["yoy_inflation_pct"], color="#7570b3", linewidth=1.5)
    plt.axhline(0, color="black", linewidth=0.8)
    plt.title("Year-on-Year Inflation in Overall CPI")
    plt.xlabel("Date")
    plt.ylabel("Year-on-year inflation (%)")
    plt.xticks(rotation=35)
    save_figure("03_yoy_inflation.png")

    plt.figure(figsize=(11, 5))
    sns.boxplot(data=prepared, x="month_number", y="index", order=MONTH_ORDER, color="#8dd3c7")
    plt.title("Seasonal Distribution of Overall CPI by Calendar Month")
    plt.xlabel("Calendar month")
    plt.ylabel("CPI index")
    plt.xticks(ticks=np.arange(12), labels=MONTH_LABELS)
    save_figure("04_seasonal_cpi_boxplot.png")

    decomposition = seasonal_decompose(prepared.set_index("date")["index"], model="additive", period=SEASONAL_PERIOD)
    figure = decomposition.plot()
    figure.set_size_inches(11, 8)
    figure.suptitle("Additive Seasonal-Trend Decomposition of Overall CPI", y=1.02)
    figure.axes[-1].set_xlabel("Date")
    figure.axes[0].set_ylabel("CPI index")
    save_figure("05_seasonal_trend_decomposition.png")

    plt.figure(figsize=(10, 5))
    plot_acf(prepared["index"], lags=min(24, len(prepared) - 1), ax=plt.gca())
    plt.title("Autocorrelation Function of Overall CPI")
    plt.xlabel("Lag (months)")
    plt.ylabel("Autocorrelation")
    save_figure("06_acf.png")

    pacf_lags = min(24, (len(prepared) // 2) - 1)
    plt.figure(figsize=(10, 5))
    plot_pacf(prepared["index"], lags=pacf_lags, method="ywm", ax=plt.gca())
    plt.title("Partial Autocorrelation Function of Overall CPI")
    plt.xlabel("Lag (months)")
    plt.ylabel("Partial autocorrelation")
    save_figure("07_pacf.png")


def split_data(prepared):
    """Create the required chronological final-12-month holdout set."""
    training = prepared.iloc[:-FORECAST_HORIZON].copy()
    testing = prepared.iloc[-FORECAST_HORIZON:].copy()
    training.to_csv(TABLES_DIR / "training_data.csv", index=False)
    testing.to_csv(TABLES_DIR / "testing_data.csv", index=False)
    print("\n--- Chronological train-test split ---")
    print(f"Training: {len(training)} observations, {training['date'].min():%Y-%m} to {training['date'].max():%Y-%m}")
    print(f"Testing: {len(testing)} observations, {testing['date'].min():%Y-%m} to {testing['date'].max():%Y-%m}")

    plt.figure(figsize=(11, 5))
    plt.plot(training["date"], training["index"], label="Training", color=COLOUR, linewidth=2)
    plt.plot(testing["date"], testing["index"], label="Testing", color="#e7298a", linewidth=2)
    plt.axvline(testing["date"].iloc[0], color="black", linestyle="--", linewidth=1, label="Split point")
    plt.title("Chronological Training and Testing Periods")
    plt.xlabel("Date")
    plt.ylabel("CPI index")
    plt.legend()
    plt.xticks(rotation=35)
    save_figure("08_training_testing_split.png")
    return training, testing


def seasonal_naive_forecast(train_values, horizon):
    """Forecast each month using its observed value one seasonal period earlier."""
    return np.resize(np.asarray(train_values.iloc[-SEASONAL_PERIOD:], dtype=float), horizon)


def holt_winters_forecast(train_values, horizon):
    model = ExponentialSmoothing(train_values, trend="add", seasonal="add", seasonal_periods=SEASONAL_PERIOD)
    return np.asarray(model.fit(optimized=True).forecast(horizon), dtype=float)


def sarima_forecast(train_values, horizon):
    model = SARIMAX(
        train_values,
        order=(1, 1, 1),
        seasonal_order=(1, 1, 1, SEASONAL_PERIOD),
        enforce_stationarity=False,
        enforce_invertibility=False,
    )
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        fitted = model.fit(disp=False)
    return np.asarray(fitted.forecast(horizon), dtype=float)


def regression_design(dates, trend_start=0):
    """Build a trend plus monthly-dummy design matrix with a stable set of columns."""
    frame = pd.DataFrame({"trend": np.arange(trend_start, trend_start + len(dates)), "month": dates.dt.month.to_numpy()})
    dummies = pd.get_dummies(frame["month"], prefix="month", dtype=float).reindex(
        columns=[f"month_{month}" for month in MONTH_ORDER], fill_value=0.0
    )
    return pd.concat([frame[["trend"]], dummies.iloc[:, 1:]], axis=1)


def regression_forecast(training, testing):
    """Fit linear trend plus 11 monthly seasonal dummies and forecast the test months."""
    model = LinearRegression()
    train_x = regression_design(training["date"], trend_start=0)
    test_x = regression_design(testing["date"], trend_start=len(training))
    model.fit(train_x, training["index"])
    return np.asarray(model.predict(test_x), dtype=float)


def calculate_metrics(actual, predicted):
    actual = np.asarray(actual, dtype=float)
    predicted = np.asarray(predicted, dtype=float)
    errors = actual - predicted
    return {
        "MAE": np.mean(np.abs(errors)),
        "RMSE": np.sqrt(np.mean(errors ** 2)),
        "MAPE (%)": np.mean(np.abs(errors / actual)) * 100,
    }


def run_forecasting_models(training, testing):
    """Fit the four member models, save comparisons, and create report-ready forecast plots."""
    # Setting the frequency explicitly avoids treating an otherwise regular
    # monthly index as an undated sequence during model fitting.
    train_values = training.set_index("date")["index"].asfreq("MS")
    forecasts = {
        "Seasonal Naive": seasonal_naive_forecast(train_values, len(testing)),
        "Holt-Winters": holt_winters_forecast(train_values, len(testing)),
        "SARIMA": sarima_forecast(train_values, len(testing)),
        "Trend + Monthly Dummies": regression_forecast(training, testing),
    }
    results = testing[["date", "index"]].rename(columns={"index": "actual_cpi"}).copy()
    metric_rows = []
    for name, forecast in forecasts.items():
        column = name.lower().replace(" ", "_").replace("+", "plus").replace("-", "_")
        results[column] = forecast
        metric_rows.append({"model": name, **calculate_metrics(testing["index"], forecast)})

        plt.figure(figsize=(11, 5))
        plt.plot(training["date"], training["index"], label="Training", color="#bdbdbd", linewidth=1.5)
        plt.plot(testing["date"], testing["index"], label="Actual test CPI", color="#e7298a", marker="o", linewidth=2)
        plt.plot(testing["date"], forecast, label=f"{name} forecast", color=COLOUR, marker="o", linestyle="--", linewidth=2)
        plt.title(f"{name}: 12-Month CPI Forecast")
        plt.xlabel("Date")
        plt.ylabel("CPI index")
        plt.legend()
        plt.xticks(rotation=35)
        safe_name = column.replace("__", "_")
        save_figure(f"09_forecast_{safe_name}.png")

    metrics = pd.DataFrame(metric_rows).sort_values("RMSE").reset_index(drop=True)
    results.to_csv(TABLES_DIR / "test_forecasts.csv", index=False)
    metrics.to_csv(TABLES_DIR / "forecast_accuracy_metrics.csv", index=False)
    print("\n--- Forecast accuracy on the final 12 months ---")
    print(metrics.to_string(index=False, float_format=lambda value: f"{value:.3f}"))
    print(f"Lowest RMSE model: {metrics.iloc[0]['model']} ({metrics.iloc[0]['RMSE']:.3f}).")

    plt.figure(figsize=(10, 5))
    sns.barplot(data=metrics, x="RMSE", y="model", hue="model", palette="Blues_d", legend=False)
    plt.title("Forecast Accuracy Comparison: Lower RMSE Is Better")
    plt.xlabel("RMSE (CPI index points)")
    plt.ylabel("Forecasting model")
    save_figure("10_forecast_accuracy_comparison.png")


def main():
    print("Forecasting Malaysia's Monthly Consumer Price Index for Low-Income Households")
    setup_output_folders()
    _, overall = load_and_validate_data()
    prepared = prepare_data(overall)
    print_interpretations(prepared)
    create_exploratory_figures(prepared)
    training, testing = split_data(prepared)
    run_forecasting_models(training, testing)
    print("\nCompleted successfully. Tables and figures are in outputs/.")


if __name__ == "__main__":
    try:
        main()
    except (ValueError, FileNotFoundError) as error:
        print(f"\nERROR: {error}", file=sys.stderr)
        sys.exit(1)
